### PR TITLE
fix(minillm): compute the discounted advantage with a stable reverse recurrence

### LINKED ISSUE #6626: MiniLLM: `_compute_advantage` discounts by absolute position, scaling every advantage by gamma^t
### Reproduction

`MiniLLMTrainer._compute_advantage` builds `gamma_pow[i] = gamma**i` from the **absolute** index `i`, then does a reverse cumsum. That computes `sum_{i>=t} gamma^i * R_i = gamma^t * A_t` instead of the documented `sum_{i>=t} gamma^(i-t) * R_i = A_t`, so every position is scaled by an extra `gamma^t`.

**No exception is raised** — with `length_normalization=False` the advantages are silently wrong (later tokens are exponentially suppressed toward zero). With `length_normalization=True` the `gamma^t` cancels between numerator and denominator, but `gamma**i` underflows to exactly `0.0` in float32, giving `0/0 = nan` on long completions.

The repro calls the method with a stub `self` because it reads only `self.gamma` and `self.length_normalization`.

```python
import types

import torch

from trl.experimental.minillm.minillm_trainer import MiniLLMTrainer

advantage = MiniLLMTrainer._compute_advantage  # real method, called with a stub `self`

# Constant reward of 1.0 at every position: rewards = teacher - student = 1 - 0 = 1
T, GAMMA = 512, 0.9
student, teacher, mask = torch.zeros(1, T), torch.ones(1, T), torch.ones(1, T)

print("BUG 1 (length_normalization=False): every position is scaled by an extra gamma^t.")
print(f"With a constant reward of 1.0, the advantage should be A_t = (1-{GAMMA}^(T-t))/(1-{GAMMA}) = 10.0")
print("at every position. Instead it decays, and the error is exactly a factor of gamma^t:\n")
out = advantage(types.SimpleNamespace(gamma=GAMMA, length_normalization=False), student, teacher, mask)[0]
print(f"{'t':>5} {'expected':>12} {'actual':>14} {'actual/expected':>17} {'gamma^t':>14}")
for t in [0, 50, 100, 200, 400]:
    expected = (1 - GAMMA ** (T - t)) / (1 - GAMMA)
    print(f"{t:>5} {expected:>12.4f} {out[t].item():>14.6e} {out[t].item() / expected:>17.6e} {GAMMA**t:>14.6e}")
print("\n^ the last two columns match: gamma_pow is indexed by the absolute i, not by (i - t).\n")

print("BUG 2 (length_normalization=True): the gamma^t factor cancels between numerator and")
print("denominator, so every position should be exactly 1.0 -- until gamma^i underflows to 0.0")
print("in float32. Past that point both sums are 0.0 and the division yields 0/0 = nan:\n")
for gamma, T2 in [(0.5, 512), (0.9, 1024), (0.95, 2048), (0.99, 16384)]:
    out = advantage(types.SimpleNamespace(gamma=gamma, length_normalization=True), torch.zeros(1, T2),
                    torch.ones(1, T2), torch.ones(1, T2))[0]
    nan_at = (~torch.isfinite(out)).nonzero()
    zero_at = (torch.pow(gamma, torch.arange(T2)) == 0.0).nonzero()
    print(f"  gamma={gamma:<5} T={T2:<5} A_0={out[0].item():<5} A_last={out[-1].item():<5} "
          f"first nan at t={str(int(nan_at[0, 0])) if len(nan_at) else 'never':<7} "
          f"first gamma^i == 0.0 at i={str(int(zero_at[0, 0])) if len(zero_at) else 'never'}")
print("\n^ the nan index always equals the underflow index. Smaller gamma underflows sooner.")
```

outputs:

```
BUG 1 (length_normalization=False): every position is scaled by an extra gamma^t.
With a constant reward of 1.0, the advantage should be A_t = (1-0.9^(T-t))/(1-0.9) = 10.0
at every position. Instead it decays, and the error is exactly a factor of gamma^t:

    t     expected         actual   actual/expected        gamma^t
    0      10.0000   9.999997e+00      9.999997e-01   1.000000e+00
   50      10.0000   5.153767e-02      5.153767e-03   5.153775e-03
  100      10.0000   2.656132e-04      2.656132e-05   2.656140e-05
  200      10.0000   7.055040e-09      7.055040e-10   7.055079e-10
  400       9.9999   4.977323e-18      4.977360e-19   4.977414e-19

^ the last two columns match: gamma_pow is indexed by the absolute i, not by (i - t).

BUG 2 (length_normalization=True): the gamma^t factor cancels between numerator and
denominator, so every position should be exactly 1.0 -- until gamma^i underflows to 0.0
in float32. Past that point both sums are 0.0 and the division yields 0/0 = nan:

  gamma=0.5   T=512   A_0=1.0   A_last=nan   first nan at t=150     first gamma^i == 0.0 at i=150
  gamma=0.9   T=1024  A_0=1.0   A_last=nan   first nan at t=987     first gamma^i == 0.0 at i=987
  gamma=0.95  T=2048  A_0=1.0   A_last=nan   first nan at t=2028    first gamma^i == 0.0 at i=2028
  gamma=0.99  T=16384 A_0=1.0   A_last=nan   first nan at t=10346   first gamma^i == 0.0 at i=10346

^ the nan index always equals the underflow index. Smaller gamma underflows sooner.
```

The referenced implementation ([LMOps `losses.py#L37-L49`](https://github.com/microsoft/LMOps/blob/e210d2c026b9958617887762400778ace81172e6/minillm/minillm/losses.py#L37-L49)) uses a backward recurrence, `last_rw = rewards[:, t] + gamma * last_rw`, which produces the relative exponent by construction and never materializes `gamma**i` — so it has neither problem. Porting that recurrence would fix both at once.

Happy to submit a PR porting the backward recurrence.

### System Info

```
- Platform: Linux-5.15.167.4-microsoft-standard-WSL2-x86_64-with-glibc2.35
- Python version: 3.12.4
- TRL version: 1.10.0.dev0+aa59202
- PyTorch version: 2.13.0
- accelerator(s): NVIDIA GeForce RTX 4090
- Transformers version: 5.14.1
- Accelerate version: 1.14.0
- Accelerate config: not found
- Datasets version: 5.0.1
- HF Hub version: 1.26.0
- bitsandbytes version: 0.50.0
- DeepSpeed version: 0.19.3
- Liger-Kernel version: 0.8.0
- PEFT version: 0.20.0
- vLLM version: not installed
```

### Checklist

- [x] I have checked that my issue isn't already filed (see [open issues](https://github.com/huggingface/trl/issues?q=is%3Aissue))
- [x] I have included my system information
- [x] Any code provided is minimal, complete, and reproducible ([more on MREs](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any code provided is properly formatted in code blocks, (no screenshot, [more on code blocks](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any traceback provided is complete

### PR DESCRIPTION
**Pull request by behroozazarkhalili**
`MiniLLMTrainer` discounted rewards by absolute token position, suppressing later advantages and producing NaNs on long normalized completions. I replaced that calculation with a matrix of relative discounts and kept the computation in float32 under autocast. The regression tests fail on main and pass here, and comparisons with a float64 reference cover varying rewards, masks, and both normalization modes.
Resolves #6626.

---

> [!NOTE]
> **Medium Risk**
> Changes core MiniLLM reverse-KL advantage math used when `gamma > 0`, so training dynamics and loss values will differ from the previous (buggy) behavior.
> 
> **Overview**
> Fixes **`MiniLLMTrainer._compute_advantage`** so advantages use **relative** discounts \(\gamma^{i-t}\) instead of weighting by absolute \(\gamma^i\) and a reverse cumsum, which wrongly scaled later tokens and could yield **NaNs** under length normalization on long completions (#6626).
> 
> The discounted sum is now a single matmul against a lower-triangular **discount matrix**, with rewards promoted to at least **float32** and **autocast disabled** for that block so mixed-precision training does not downcast the matmul to bf16/fp16.
> 
> Adds **`TestMiniLLMComputeAdvantage`** (stub trainer, no model): closed-form constant-reward check, long-sequence finiteness, float64 reference with masks for both normalization modes, and autocast dtype stability.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 09e50538652fd2c66c4e47e8b3d7e725d9b0d3c1. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: trl/experimental/minillm/minillm_trainer.py [hunk 1]
```diff
@@ -330,15 +330,24 @@ def _compute_advantage(
...
+            # advantages_t = sum_{i>=t} gamma^(i-t) R_i, computed by the reverse recurrence
+            # a_t = R_t + gamma * a_{t+1} (a_T = 0). The previous implementation weighted by gamma^i (the
+            # absolute index), which scaled every advantage by an extra gamma^t and underflowed to 0.0 in
+            # float32 on long completions, giving 0/0 = nan under length normalization. The recurrence never
+            # materializes gamma^i, so it is exact and stable at any length. See issue #6626.
+            advantages = torch.zeros_like(rewards)
+            running_advantage = torch.zeros_like(rewards[:, 0])
+            for t in range(response_length - 1, -1, -1):
+                running_advantage = rewards[:, t] + self.gamma * running_advantage
+                advantages[:, t] = running_advantage
```
[the remark this answers is not in the document]
Comment (behroozazarkhalili) [reply trl/experimental/minillm/minillm_trainer.py]: I believe this one is a false alarm: the recurrence does preserve autograd. The in-place indexed write `advantages[:, t] = running_advantage` is a differentiable op (`CopySlices`), so writing a grad-requiring value into the `zeros_like` buffer connects it to the graph rather than detaching it.

Verified empirically with `student_log_probs_on_labels` requiring grad:

```python
student = torch.zeros(1, 16, requires_grad=True)
teacher = torch.ones(1, 16)
adv = MiniLLMTrainer._compute_advantage(stub, student, teacher)
adv.sum().backward()
```

Result:

```
length_normalization=False : adv.requires_grad=True, grad_fn=CopySlices,  student.grad -> 16/16 non-zero
length_normalization=True  : adv.requires_grad=True, grad_fn=DivBackward0, student.grad -> 16/16 non-zero
```

This matches the previous `flip/cumsum/flip` path (`grad_fn=FlipBackward0`, also 16/16 non-zero), so the gradient connectivity the trainer relies on is intact, and `Tests (experimental)` is green on this PR. I am leaving the code as is, but happy to switch the buffer write to `torch.stack([...])` if a form that makes the autograd-safety obvious at a glance is preferred. cc @qgallouedec

[the remark this answers is not in the document]
Comment (behroozazarkhalili) [reply trl/experimental/minillm/minillm_trainer.py]: Resolving this on the basis of the measurement above: the indexed write is a differentiable `CopySlices`, so the returned `advantages` stays connected to the graph, verified as `requires_grad=True` with 16/16 non-zero student gradients on both the normalized and unnormalized paths, matching the previous `flip/cumsum/flip` behaviour. Happy to reopen if anyone reads it differently, and the `torch.stack` alternative is still on the table if a form that is obviously safe at a glance is preferred.

**Comment by LiRunGuo** [comment]
Thanks for the fix — the reverse-recurrence approach (advantage[t] = rewards[t] + γ·advantage[t+1], and the same recurrence for the length-normalization denominator) matches both the docstring formula and the reference implementation this method cites. I independently derived the same design while reproducing #6626, and since you already cover the fix here I'm not opening a separate PR. Instead, a few regression scenarios that I verified behave badly on the pre-fix code would strengthen the test coverage beyond constant rewards:

1. **Random per-position rewards vs. a naive summation of the documented formula** (both `length_normalization` modes). Constant rewards are a special case: with length normalization the numerator and denominator are identical, so subtle indexing errors can cancel out and stay invisible.
2. **Trailing masked rows** (right-padded, e.g. rows ending at different positions). This exercises the `mask < 0.5 → 1e-4` denominator path and ensures masked positions never leak masked rewards into the discounted sum.
3. **`gamma == 0` passthrough** (advantages must equal the raw rewards unchanged).
4. **Long-sequence underflow at the parameters from the issue's own repro** — γ=0.9, length 2048, constant reward, length-normalized: assert `torch.isfinite(...).all()` and advantages ≈ 1.0 everywhere. Your γ=0.5/T=512 case already covers the underflow mechanism; this pins the exact configuration reported in #6626 too.

On the pre-fix code, scenarios 1 (non-normalized), 2/4 and the "discounted from position t" closed-form check fail; post-fix all pass. Happy to share the exact test class if you'd like to fold any of it into this PR (e.g. as a follow-up gist or snippet) — just say the word.

(Note for transparency: this review comment was drafted with AI assistance.)
